sound/soc/sof/ipc4-pcm.c

```diff
@@ -941,16 +960,29 @@ static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
 	 * Firmware calculates correct stream_start_offset for all cases including above two.
 	 * Driver subtracts stream_start_offset from dai dma position to get accurate one
 	 */
-	tmp_ptr -= time_info->stream_start_offset;
 
-	/* Calculate the delay taking into account that both pointer can wrap */
-	div64_u64_rem(tmp_ptr, substream->runtime->boundary, &tmp_ptr);
+	/*
+	 * If the dai position has passed the start offset it means that the
+	 * audio left the DSP and we need to normalize the dai position
+	 * Otherwise the first audio frames is still within the DSP, we should
+	 * adjust the host side position to accurately report the initial delay
+	 */
+	if (tmp_ptr >= time_info->stream_start_offset) {
+		tmp_ptr -= time_info->stream_start_offset;
+	} else {
+		host_ptr += time_info->stream_start_offset - tmp_ptr;
+		tmp_ptr = 0;
+	}
+
+	/* Wrap the dai position at the boundary where the host position wraps */
+	div64_u64_rem(tmp_ptr, time_info->boundary, &tmp_ptr);
```

Comment:
Rest of this patch looks good and I can follow the explanations, but I fail to understand how the connection between time_info->boundary and substream->runtime->boundary (still used on L989) is made. How do we know host position wraps at "time_info->boundary"? And shouldn't we use time_info->boundary on L989?

Reply:
@kv2019i, oh yes, you are right, I need to swap that as well, thanks for finding it!, it would have taken few decades to hit the wrap ;)